### ISSUE TITLE
Ability to do line-level customisation (suppression) of conformance rules

### ISSUE THREAD

**Issue by supersteves** [google/closure-compiler#1434]
I'm experimenting with conformance rules but have found I need to suppress a few specific violations, and would rather do this on individual lines than for an entire file.

I haven't found how to do this, other than rolling my own solution (adding support for using ```@suppress {conformanceViolations}``` as in https://github.com/google/closure-compiler/pull/1199, or by writing my own warnings guard which read some proprietary hints file).

Quite often this is because there is some compiler oddity, perhaps a bug which I would file, but in the interim would want to maintain my project's "clean slate" warnings policy.

Am I missing something and is this already possible? If not please could you reconsider the very low risk single line changing PR https://github.com/google/closure-compiler/pull/1199. Currently I have to maintain a custom fork for this change.

**Comment by ChadKillingsworth** [comment]
Suppression in general is only supported on functions or the fileoverview block. Per statement suppressions are considered ambiguous.

**Comment by supersteves** [comment]
Ah yes, I forgot. Updated.

**Comment by concavelenz** [comment]
Yes, as stated this is deliberate.   Conformance is used for security and other issues and the only way we could support this would be to add a "source suppressible" attribute to the conformance rule definition and use a separate diagnostic type and this could be added as a conformance rule.

I think this would work for you as you could then add this to your rules and allow you to suppress on a per-function basis.

thoughts?

**Comment by supersteves** [comment]
Do you mean that you don't want to allow most rules to be "stealthily suppressed"? 

In my case I am looking to catch programmer errors - I was suggested to use conformance rules rather than the old "unknown types" check. Avoiding compiler bugs and (e.g. [O/N]TI) limitations are genuine case to permit suppression, given your restriction.

If there was such a restriction on its use, is ```@suppress {conformanceViolations}``` an acceptable way of marking suppression?

Should I put a CompilerOption in for it, ```allowSuppressConformanceRules=false``` by default? Or would you prefer ```suppressible``` inserted at https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/conformance.proto#L127 ?

**Comment by concavelenz** [comment]
Now that I've thought about this a bit, I have a better idea: give the conformance rule a "level" property:  lint (or info), warning, error.  Given the uses folks for putting the framework to this would be a lot more flexible. "lint" would never be promoted to error, "warning" could but be suppressible via whitelist or annotation, "error" would only be suppressible via the conformance rule configuration itself.

This is a bit more work, let me see if the security guys want to take it on.

**Comment by supersteves** [comment]
Makes sense. Allow the conformance rule to indicate its purpose, and implicitly whether it is suppressible using existing whitelist or planned ```@suppress```. 

Not sure about the distinction between lint and warning, though. Both can be suppressed, both default to warning-like reporting. While I can see a benefit in preventing people from demoting errors to warnings for security, I don't see any reason to prevent people elevating warnings to errors. 

But it makes sense to have a separate classification "lint" even if it's equivalent to warning in effect. Although linters tend to have detection of more serious issues, the word probably means fluff that sullies the look of the code without much risk of affecting runtime behaviour. I.e. mostly code style. For an OS project with many contributors you might want to elevate certain linting to errors such that CI might prevent PR integration in the wrong code style, for example.
